tls/s2n_handshake_io.c

```diff
@@ -1040,27 +1015,41 @@ int s2n_negotiate(struct s2n_connection *conn, s2n_blocked_status * blocked)
         this = 'C';
     }
 
-    while (ACTIVE_STATE(conn).writer != 'B') {
-        /* Flush any pending I/O or alert messages */
-        GUARD(s2n_flush(conn, blocked));
+    /* If we were blocked reading a record, call the handler right away. Don't try to read more data. */
```

Comment:
Callers are supposed to trigger s2n_negotiate in a non_blocking way, which means only when session data is available, it would be called again. Although we don't know how exactly caller would do, since you have `client_hello_finished` flag, it should be safe to call `s2n_establish_session` twice or more.